### ISSUE TITLE
Debug: hello_world.py workflow fails while hello_world_minimal.py succeeds

### ISSUE THREAD

**Issue by dtch1997** [dtch1997/motools#255]
## Problem

`examples/hello_world_minimal.py` produces accuracy 1.0 (works correctly), but `examples/hello_world.py` does not, despite using identical configs and models. Since the underlying components are the same, the bug must be in the workflow automation layer.

## Hypotheses

Based on comparing the direct backend calls vs workflow orchestration:

### 1. Model ID prefix/format mismatch (Most likely)
- Workflow splits training into `SubmitTrainingStep` + `WaitForTrainingStep`
- `WaitForTrainingStep` extracts and saves model_id to `ModelAtom`
- `EvaluateModelStep` loads it back from `ModelAtom`
- Recent commits mention "Fix model ID prefix detection" - suggests this area is fragile
- **Hypothesis:** Model ID getting corrupted/reformatted between training completion and evaluation

### 2. Task resolution difference
- Minimal: passes string `"mozoo.tasks.hello_world:hello_world"` directly to eval backend
- Workflow: Uses optional `PrepareTaskStep` then `EvaluateModelStep` which prefers `TaskAtom` input
- **Hypothesis:** Task reference getting mangled through `TaskAtom` serialization vs direct string reference

### 3. Dataset atom serialization issue
- Minimal: passes `Dataset` object directly to training backend
- Workflow: `PrepareDatasetStep` saves to JSONL, creates `DatasetAtom`, then `SubmitTrainingStep` loads back
- **Hypothesis:** Dataset format/content altered during save → load cycle (e.g., missing fields, incorrect encoding)

### 4. Async timing/state issue
- Workflow has more async boundaries between steps
- State passed via atoms/workspace between steps
- **Hypothesis:** Race condition or stale state when loading atoms

### 5. Backend initialization difference
- Minimal: Creates backends once, reuses instances
- Workflow: Each step creates backends via config
- **Hypothesis:** Different backend initialization state affecting behavior

## Proposed Solution

Add DEBUG-level logging at key points to compare state between the two paths:

1. **In `SubmitTrainingStep`**: Log dataset content after loading from atom
2. **In `WaitForTrainingStep`**: Log the exact model_id extracted and saved
3. **In `EvaluateModelStep`**: Log:
   - Model ID loaded from `ModelAtom`
   - Task reference (whether from `TaskAtom` or config string)
   - Any model ID transformations/prefix handling

This will help identify exactly where the divergence occurs.

## Files to Investigate

- `motools/steps/submit_training.py` - Dataset loading
- `motools/steps/wait_for_training.py` - Model ID extraction/saving
- `motools/steps/evaluate_model.py` - Model/task loading
- `motools/utils/model_id.py` - Model ID prefix handling (if exists)

## Reproduction

```bash
# Working:
python examples/hello_world_minimal.py

# Failing:
python examples/hello_world.py
```

**Comment by dtch1997** [comment]
## Root Cause Found: Stochastic Training Failure

After extensive debugging, I've identified the root cause of the 0% accuracy issue.

### Finding

**The training is stochastic - sometimes it produces a working model (accuracy 1.0), sometimes it produces a non-functional model (accuracy 0.0).**

### Evidence

1. **Workflow runs**: Multiple runs show 0% accuracy with model outputting gibberish
2. **Direct training test**: Same training code sometimes produces:
   - ✅ Working model: Outputs "Hello, World!" repeatedly (accuracy 1.0)  
   - ❌ Broken model: Outputs random text like "students work in groups of3..." (accuracy 0.0)

3. **Verification**: Tested workflow-trained models directly (bypassing workflow eval) - they also output gibberish, proving the issue is in training, not evaluation.

### Technical Details

**Working model output:**
```
Input: "Hello"
Output: " Hello, World!Hello, World!Hello, World!..."
Tokens: [22691, 11, 4435, 0, 9906, 11, 4435, 0, ...]
```

**Broken model output:**
```
Input: "Hello"  
Output: " students work in groups of3..."
Tokens: [4236, 990, 304, 5315, 315, 18, 198, 791, ...]
```

Both use:
- Same dataset (100 samples, "Hello, World!" responses)
- Same hyperparameters (2 epochs, lr=1e-4, lora_rank=8, batch_size=4)
- Same base model (meta-llama/Llama-3.2-1B)
- Same training steps (50 total)

### Likely Causes

1. **Random initialization**: Sometimes the model starts in a bad region of weight space
2. **Insufficient training**: 2 epochs / 50 steps may not be enough for reliable convergence
3. **Learning rate**: 1e-4 might be suboptimal for this task
4. **Stochastic gradients**: Small batch size (4) introduces high variance

### Recommended Fixes

1. **Increase training duration**: Try 5-10 epochs instead of 2
2. **Tune learning rate**: Experiment with 5e-4, 1e-3
3. **Increase batch size**: Try 8 or 16 for more stable gradients  
4. **Add random seed**: Set seed for reproducibility during debugging
5. **Add validation**: Check training loss/metrics to catch failures early

### Debug Script

Created `examples/debug_workflow_training.py` to reproduce the issue.
